src/System.Management.Automation/engine/lang/parserutils.cs

```diff
@@ -681,24 +696,30 @@ private static object SplitWithPredicate(ExecutionContext context, IScriptExtent
                         // If this delimiter is at the end of the string,
                         // add an empty string to denote the item "after"
                         // it.
-                        if (strIndex == (item.Length - 1))
+                        if (cursor == (item.Length - 1))
                         {
                             split.Add("");
                         }
                     }
                     else
                     {
-                        buf.Append(item[strIndex]);
+                        if(rightToLeft)
+                            buf.Insert(0, item[strIndex]);
+                        else
+                            buf.Append(item[strIndex]);
                     }
                 }
 
                 // Add any remainder, if we're under the limit.
                 if (buf.Length > 0 &&
-                    (limit <= 0 || split.Count < limit))
+                    (limit == 0 || split.Count < limit))
                 {
                     split.Add(buf.ToString());
                 }
 
+                if(rightToLeft)
+                    split.Reverse();
```

Comment:
Can we exclude `reverse`?

Reply:
Remove the `List<string>.Reverse()` call for performance reasons or...? 

Removing it would funtionally result in the substrings for each input being returned in the wrong order, ie:

```
PS> 'a b c d' -split {$_ -eq ' '},-3
d
c
a b
```

Rather than the expected:

```
PS> 'a b c d' -split {$_ -eq ' '},-3
a b
c
d
```

I see two possible alternatives: 
1. Change all invocations of `split.Add()` to `split.Insert()` conditional on `rightToLeft`
2. Implement two completely separate routines for negative and positive `limit` values, duplicating the code slightly (but keeping performance overhead at a minimum)

Reply:
Yes, my thoughts was about performance. I don't know what case is better. I guess we would use only one routine. At a minimum, we have to start with this.

Reply:
Splitting it into 2 loops would mean getting rid of at least one non-local variable assignment per input item and a number of branch points, the intent (splitting from the end) would also be clearer. Any thoughts @lzybkr ?

Reply:
`List.Reverse` is not expensive, it swaps in place, see the [implementation](http://source.dot.net/#System.Private.CoreLib/src/System/Array.cs,b242f3290263f0f2) (List is really an Array).

Insert would be more expensive, shifting everything on each insert instead of swapping once like in Reverse.

Reply:
We could do reverse copy in next line in `ExtendList`.
By the way, what are we doing this recopy?

Reply:
Up.